### PR TITLE
executor: consider collations when calculating hash values in `HashJoin`

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb/issues/14659.
Consider collations when calculating hash values in `HashJoin`.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: util/codec/codec_test.go [hunk 1]
```diff
@@ -1216,3 +1216,96 @@ func (s *testCodecSuite) TestHashChunkColumns(c *C) {
...
+		h1 := crc32.NewIEEE()
+		h2 := crc32.NewIEEE()
+		c.Assert(HashChunkRow(sc, h1, chk1.GetRow(i), tps, cols, buf), IsNil)
+		c.Assert(HashChunkRow(sc, h2, chk2.GetRow(i), tps, cols, buf), IsNil)
+		c.Assert(h1.Sum32(), Not(Equals), h2.Sum32())
+		h1.Reset()
+		h2.Reset()
+	}
+
+	tp.Collate = "utf8_general_ci"
```
Comment (bb7133) [inline util/codec/codec_test.go]: The current problem is, for now we have to call `collate.SetNewCollationEnabled(true)` to get the "real" general_ci collator. Maybe we should add an interface for the unit tests.

Comment (qw4990) [reply util/codec/codec_test.go]: I add the call `collate.SetNewCollationEnabledForTest(true)` in each tests about collations now.
And all these tests are processed by the `testCollationSuite`, which is a `SerialSuites`.

**Pull request merged**